src/Database/LSMTree/Internal/RunReaders.hs

```diff
@@ -186,12 +216,18 @@ nextReadCtx ::
      HasFS IO h
   -> HasBlockIO IO h
   -> ReaderNumber
-  -> RunReader IO (FS.Handle h)
+  -> Reader IO (FS.Handle h)
   -> IO (Maybe (ReadCtx (FS.Handle h)))
-nextReadCtx fs hbio readCtxNumber readCtxReader = do
-    res <- Reader.next fs hbio readCtxReader
-    case res of
-      Reader.Empty -> do
-        return Nothing
-      Reader.ReadEntry readCtxHeadKey readCtxHeadEntry ->
-        return (Just ReadCtx {..})
+nextReadCtx fs hbio readCtxNumber readCtxReader =
+    case readCtxReader of
+      ReadRun r -> flip fmap (Reader.next fs hbio r) $ \case
```

Comment:
But then I need to `pure/return`. :shrug: 
Some people use `<&>`, but I don't think we do anywhere. So this seemed like a reasonable solution, especially since the `ReadBuffer` version below looks pretty similar, applying a pure function as well.

Reply:
I overlooked the need for `pure`/`return`. Arguably, `<&>` would be slightly better